tiledbsoma 1.8.0 pre-check

johnkerl opened the pull request:
Following our [established procedure](https://github.com/single-cell-data/TileDB-SOMA/wiki/Branches-and-releases)

See also #93 which addresses issues which have happened since 1.7.2 -- things involving our dependencies -- which have  nothing to do with 1.8.0

See also https://github.com/single-cell-data/TileDB-SOMA/issues/2219 for tracking.

johnkerl commented:
[sc-41030]

jdblischak commented:
Currently failing with similar conda solver errors as we previously identified in https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/89#issuecomment-1965013717

We are getting conda solver errors in multiple feedstocks, so there must be some systemic problem. I'm hoping that this [commit](https://github.com/conda-forge/tiledb-feedstock/commit/a040ee47301cdd40495ce065cb17fe9d88374b9b) merged earlier this morning (which migrates to libgoogle-cloud 2.21) will help ease the solver problems 🤞

johnkerl commented:
> I'm hoping that this [commit](https://github.com/conda-forge/tiledb-feedstock/commit/a040ee47301cdd40495ce065cb17fe9d88374b9b) merged earlier this morning (which migrates to libgoogle-cloud 2.21) will help ease the solver problems 🤞

@jdblischak are we waiting for some sort of propagation?

jdblischak commented:
> are we waiting for some sort of propagation?

Yes, I'm hoping that maybe the new tiledb 2.20.1 binaries that were built against libgoogle-cloud 2.21 might help with the current solver issue. They were uploaded more than 2 hours ago, so should be available now

https://anaconda.org/conda-forge/tiledb/files?version=2.20.1

(note that the build numbers are weird (two different sets with build number 1) because of all the merge commits we've done recently in tiledb-feedstock. Fortunately the migration to libgoogle-cloud 2.21 also changed the build hash, so new binaries were uploaded)

jdblischak commented:
The solver error from the [latest run](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=38112&view=logs&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=986b1512-c876-5f92-0d81-ba851554a0a3) was decently informative. We know it has to do with libabseil, libcurl, and libgoogle-cloud-storage

```
    Could not solve for environment specs
    The following packages are incompatible
    ├─ libabseil ==20240116.1 cxx17_h59595ed_1 is requested and can be installed;
    ├─ libcurl ==8.5.0 hca28451_0 is requested and can be installed;
    ├─ libgoogle-cloud-storage ==2.21.0 hc7a4891_2 is requested and can be installed;
    └─ libtiledbsoma ==1.8.0 hd763302_0 is not installable because it requires
       └─ tiledb >=2.20.1,<2.21.0a0  but there are no viable options
          ├─ tiledb 2.20.1 would require
          │  └─ libcurl >=7.88.1,<8.0a0 , which conflicts with any installable versions previously reported;
          ├─ tiledb 2.20.1 would require
          │  └─ libgoogle-cloud-storage >=2.17.0,<2.18.0a0 , which conflicts with any installable versions previously reported;
          └─ tiledb 2.20.1 would require
             └─ libabseil >=20230802.1,<20230803.0a0 , which conflicts with any installable versions previously reported.
```

johnkerl commented:
https://anaconda.org/conda-forge/google-cloud-cpp/files updated 4 days ago ... 🤔

johnkerl commented:
Weird thing is 2.17 doesn't exist _at all_ at https://anaconda.org/conda-forge/google-cloud-cpp/files ... 🤔

jdblischak commented:
https://anaconda.org/conda-forge/libgoogle-cloud-storage/files?version=2.17.0

johnkerl commented:
OK I may be mixing `google-cloud-cpp` and `google-cloud-storage` ... OTOH from the last raw log:

```
johnkerl@exade[prod][][~]$ grep -C1 google-cloud Desktop/raw.txt
2024-02-28T22:31:52.2156877Z     libgomp:                  13.2.0-h807b86a_5           conda-forge
2024-02-28T22:31:52.2157237Z     libgoogle-cloud:          2.21.0-h31df0ca_1           conda-forge
2024-02-28T22:31:52.2157609Z     libgoogle-cloud-storage:  2.21.0-hc7a4891_1           conda-forge
2024-02-28T22:31:52.2157976Z     libgrpc:                  1.60.1-h74775cd_0           conda-forge
--
2024-02-28T22:34:23.3632616Z     libgomp:                   13.2.0-h807b86a_5           conda-forge
2024-02-28T22:34:23.3633104Z     libgoogle-cloud:           2.17.0-h31df0ca_2           conda-forge
2024-02-28T22:34:23.3634545Z     libgoogle-cloud-storage:   2.17.0-hc7a4891_2           conda-forge
2024-02-28T22:34:23.3635111Z     libgrpc:                   1.60.1-h74775cd_0           conda-forge
--
2024-02-28T22:36:32.5382008Z     libgomp:                   13.2.0-h807b86a_5           conda-forge
2024-02-28T22:36:32.5382437Z     libgoogle-cloud:           2.17.0-h31df0ca_2           conda-forge
2024-02-28T22:36:32.5382872Z     libgoogle-cloud-storage:   2.17.0-hc7a4891_2           conda-forge
2024-02-28T22:36:32.5383298Z     libgrpc:                   1.60.1-h74775cd_0           conda-forge
--
2024-02-28T22:38:44.8621716Z     ├─ libcurl ==8.5.0 hca28451_0 is requested and can be installed;
2024-02-28T22:38:44.8622037Z     ├─ libgoogle-cloud-storage ==2.21.0 hc7a4891_2 is requested and can be installed;
2024-02-28T22:38:44.8622353Z     └─ libtiledbsoma ==1.8.0 hd763302_0 is not installable because it requires
--
2024-02-28T22:38:44.8623815Z           ├─ tiledb 2.20.1 would require
2024-02-28T22:38:44.8624167Z           │  └─ libgoogle-cloud-storage >=2.17.0,<2.18.0a0 , which conflicts with any installable versions previously reported;
2024-02-28T22:38:44.8636944Z           └─ tiledb 2.20.1 would require
```

🤔

johnkerl commented:
[sc-42200]

ihnorton commented on recipe/meta.yaml:
```diff
@@ -90,6 +91,7 @@ outputs:
         - pyarrow <13.0 # [osx]
         - pyarrow       # [not osx]
         - pyarrow-hotfix
+        - libgoogle-cloud 2.12.*
```
Isn't this supposed to be 2.21?

ihnorton replied:
(actually, why are we pinning this at all?)

johnkerl replied:
It is not

See Slack

johnkerl replied:
I'll add some details here momentarily

jdblischak replied:
> (actually, why are we pinning this at all?)

It's a hacky workaround we are applying to unblock tiledbsoma. Originally discovered in #89, this solver error came out of nowhere. I suspect it has to do with the recent conda-forge switch from google-cloud-cpp to libgoogle-cloud, but I've been too busy with the libcurl 7 stuff to figure out exactly what is going on.

ihnorton replied:
Ok, carry on - thanks.

johnkerl replied:
For this PR 2.12 doesn't work although it did for #93:
https://gist.github.com/johnkerl/7812fb057db778d8bfb17b424dec6ead

Trying 2.17 locally ... 🤞

jdblischak replied:
libgoogle-cloud 2.17 fixed the conda solver error. However, only the linux-64 build is green

jdblischak commented:
The [osx-64 build](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=38142&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b) is now failing because the r-tiledbsoma compilation is failing with a bunch of macOS SDK errors. Maybe we need to adjust `MACOSX_SDK_VERSION` and/or `MACOSX_DEPLOYMENT_TARGET`?

https://github.com/TileDB-Inc/tiledbsoma-feedstock/blob/6663e5111ccd0a74dd5e4f01781918f4bf52879d/recipe/conda_build_config.yaml#L1-L4

Here is the final part of the errors:

```
/Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_build_env/bin/../include/c++/v1/__filesystem/path.h:780:25: note: 'operator basic_string' has been explicitly marked unavailable here
  _LIBCPP_HIDE_FROM_ABI operator string_type() const { return __pn_; }
                        ^
In file included from RcppExports.cpp:4:
In file included from ./../inst/include/tiledbsoma_types.h:21:
In file included from /Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/include/tiledbsoma/tiledbsoma:51:
/Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/include/tiledbsoma/soma/soma_collection.h:105:44: error: 'path' is unavailable: introduced in macOS 10.15 - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
              std::string(std::filesystem::path(uri).filename()),  // group name
                                           ^
/Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_build_env/bin/../include/c++/v1/__filesystem/path.h:442:24: note: 'path' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS path {
                       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [/Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/etc/Makeconf:200: RcppExports.o] Error 1
ERROR: compilation failed for package ‘tiledbsoma’
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1709247111523/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/tiledbsoma’
```

In the [osx-arm64 build](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=38142&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba), r-tiledbsoma compiles fine. The failure is due to tiledbsoma-py being unable to find libtiledbsoma.so in the conda env, and then failing to build it from source:

```
running build_ext
Checking: /Users/runner/miniforge3/conda-bld/tiledbsoma_1709247012541/work/dist/lib/libtiledbsoma.dylib exists: False
dlopen(libtiledbsoma.dylib, 6): image not found
Build with TileDB: /Users/runner/miniforge3/conda-bld/tiledbsoma_1709247012541/work/dist
Building Release build
./bld: line 88: cmake: command not found
```

My guess is that this might be due to https://github.com/single-cell-data/TileDB-SOMA/pull/1937 since it updated how tiledbsoma-py interacts with an existing libtiledbsoma.so

jdblischak commented:
From the [nightly feedstock builds](https://dev.azure.com/TileDB-Inc/CI/_build?definitionId=43&_a=summary&repositoryFilter=39&branchFilter=8732%2C8732%2C8732%2C8732), the R compilation error on osx-64 started on [Feb 12th](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=37882&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b). And the osx-arm64 builds started failing on [Feb 16th](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=37958&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba), so I am more confident that it was caused by https://github.com/single-cell-data/TileDB-SOMA/pull/1937, which was merged earlier that day in https://github.com/single-cell-data/TileDB-SOMA/commit/f4999d8ea89f7fcfb710380e64f3572acd01d5ca

johnkerl commented:
Back in draft -- please ignore until I mark this ready for review again -- I'm trying to figure out (mainly via local builds) what worked about #93 that didn't work here ...

johnkerl commented:
Oops, I missed comment
https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/91#issuecomment-1972361731
... 👀

johnkerl commented:
@jdblischak FYI `- libgoogle-cloud 2.17.*` no longer solves -- this happened today -- I've removed that line

I've confirmed this locally (for linux) as well

johnkerl commented:
@jdblischak @ihnorton linux is once again green.

OSX is not; and I've switched from 10.14 to 10.15 in an attempt to rid errors of the form
```
2024-03-01T22:51:51.7619640Z /Users/runner/miniforge3/conda-bld/tiledbsoma_1709332344079/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/include/tiledbsoma/soma/soma_sparse_ndarray.h:121:54: error: '~path' is unavailable: introduced in macOS 10.15 - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
2024-03-01T22:51:51.7620440Z               std::string(std::filesystem::path(uri).filename()),  // array name
2024-03-01T22:51:51.7620860Z                                                      ^
2024-03-01T22:51:51.7621620Z /Users/runner/miniforge3/conda-bld/tiledbsoma_1709332344079/_build_env/bin/../include/c++/v1/__filesystem/path.h:505:3: note: '~path' has been explicitly marked unavailable here
2024-03-01T22:51:51.7622230Z   ~path() = default;
2024-03-01T22:51:51.7622590Z   ^
```

Note that these no longer say anything about 10.14 (as they did before) so we may be looking at a genuinely deleted constructor ... 🤔

johnkerl commented:
... raw logs reveal 10.14 still appearing in log output. Trying a rerender ...

johnkerl commented:
https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/91/commits/d416ebc97e31711753b21d280d547389f53ad7a4 🤞

johnkerl commented:
See also https://github.com/single-cell-data/TileDB-SOMA/issues/2210

johnkerl commented:
CI is green; this PR has done its job per our [established procedure](https://github.com/single-cell-data/TileDB-SOMA/wiki/Branches-and-releases)

See also notes on https://github.com/single-cell-data/TileDB-SOMA/issues/2210 that the `cmake` dependency in the `build` line here is a workaround not a fix -- build-from-source should not be happening in the first place

johnkerl commented:
More advice from @jdblischak

jdblischak commented:
The `--no-build-isolation` flag is working as intended. `pip` is no longer installing the build deps

xref: https://github.com/TileDB-Inc/tiledbsoma-feedstock/pull/95#issuecomment-1981596862

```
/home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/python -m pip install --no-deps --no-build-isolation . -vv
Using pip 24.0 from /home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.10/site-packages/pip (python 3.10)
Non-user install because user site-packages disabled
Created temporary directory: /tmp/pip-build-tracker-_creqxao
Initialized build tracking at /tmp/pip-build-tracker-_creqxao
Created build tracker: /tmp/pip-build-tracker-_creqxao
Entered build tracker: /tmp/pip-build-tracker-_creqxao
Created temporary directory: /tmp/pip-install-qfsfg8yk
Created temporary directory: /tmp/pip-ephem-wheel-cache-ly63dwyh
Processing /home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/work/apis/python
  Added file:///home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/work/apis/python to build tracker '/tmp/pip-build-tracker-_creqxao'
  Created temporary directory: /tmp/pip-modern-metadata-3e9ncd7i
  Preparing metadata (pyproject.toml): started
  Running command Preparing metadata (pyproject.toml)
  running dist_info
  creating /tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info
  writing /tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/PKG-INFO
  writing dependency_links to /tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/dependency_links.txt
  writing requirements to /tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/requires.txt
  writing top-level names to /tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/top_level.txt
  writing manifest file '/tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no files found matching 'py.typed'
  no previously-included directories found matching 'dist_links/libtiledbsoma/test/__pycache__'
  writing manifest file '/tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma.egg-info/SOURCES.txt'
  creating '/tmp/pip-modern-metadata-3e9ncd7i/tiledbsoma-1.5.0rc0.post226.dev267876719.dist-info'
  Preparing metadata (pyproject.toml): finished with status 'done'
  Source in /home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/work/apis/python has version 1.5.0rc0.post226.dev267876719, which satisfies requirement tiledbsoma==1.5.0rc0.post226.dev267876719 from file:///home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/work/apis/python
  Removed tiledbsoma==1.5.0rc0.post226.dev267876719 from file:///home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/work/apis/python from build tracker '/tmp/pip-build-tracker-_creqxao'
Created temporary directory: /tmp/pip-unpack-q9ol_44z
Building wheels for collected packages: tiledbsoma
  Created temporary directory: /tmp/pip-wheel-shq_ssco
  Destination directory: /tmp/pip-wheel-shq_ssco
  Building wheel for tiledbsoma (pyproject.toml): started
  Running command Building wheel for tiledbsoma (pyproject.toml)
  running bdist_wheel
  Checking: /home/conda/feedstock_root/build_artifacts/tiledbsoma_1709752782816/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/libtiledbsoma.so exists: True
  running build
  running build_py
```
